hermes-client/src/main/java/pl/allegro/tech/hermes/client/HermesClient.java

```diff
@@ -1,42 +1,67 @@
+
 import java.net.URI;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.CompletionStage;
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicInteger;
-import java.util.function.Function;
 import java.util.function.Predicate;
 
-import static java.util.concurrent.CompletableFuture.completedFuture;
-import static java.util.stream.IntStream.range;
 import static pl.allegro.tech.hermes.client.HermesMessage.hermesMessage;
 
 public class HermesClient {
 
     private final HermesSender sender;
     private final String uri;
     private final Map<String, String> defaultHeaders;
-    private final int retries;
-    private final Predicate<HermesResponse> retryCondition;
     private final AtomicInteger currentlySending = new AtomicInteger(0);
+    private final RetryPolicy retryPolicy;
+    private final ScheduledExecutorService scheduler;
     private volatile boolean shutdown = false;
 
     HermesClient(HermesSender sender,
                  URI uri,
                  Map<String, String> defaultHeaders,
                  int retries,
-                 Predicate<HermesResponse> retryCondition) {
+                 Predicate<HermesResponse> retryCondition,
+                 long retrySleepInMillis,
+                 long maxRetrySleepInMillis) {
+        this(sender, uri, defaultHeaders, retries, retryCondition, retrySleepInMillis,
+                maxRetrySleepInMillis, Executors.newSingleThreadScheduledExecutor());
+    }
+
+    HermesClient(HermesSender sender,
+                 URI uri,
+                 Map<String, String> defaultHeaders,
+                 int retries,
+                 Predicate<HermesResponse> retryCondition,
+                 long retrySleepInMillis,
+                 long maxRetrySleepInMillis,
+                 ScheduledExecutorService scheduler) {
         this.sender = sender;
         this.uri = createUri(uri);
         this.defaultHeaders = Collections.unmodifiableMap(new HashMap<>(defaultHeaders));
-        this.retries = retries;
-        this.retryCondition = retryCondition;
+        this.retryPolicy = createRetryPolicy(retries, retryCondition, retrySleepInMillis, maxRetrySleepInMillis);
+        this.scheduler = scheduler;
+    }
+
+    private RetryPolicy createRetryPolicy(int retries, Predicate<HermesResponse> retryCondition,
+                                          long retrySleepInMillis, long maxRetrySleepInMillis) {
+        RetryPolicy retryPolicy = new RetryPolicy().withMaxRetries(retries)
+                .retryIf(retryCondition::test);
+        if (retrySleepInMillis > 0) {
+            retryPolicy.withBackoff(retrySleepInMillis, maxRetrySleepInMillis, TimeUnit.MILLISECONDS);
```

Comment:
Unfortunately no. But such a test would require to expose too many internals and hardwire to implementation:

    def "should sleep between retries when retry sleep is provided"() {
        given:
            ScheduledExecutorService scheduler = Spy(ScheduledThreadPoolExecutor, constructorArgs: [1])
            HermesClient client = hermesClient({uri, msg -> statusFuture(500)})
                    .withRetries(2)
                    .withRetrySleep(10)
                    .withScheduler(scheduler)
                    .build()

        when:
            client.publish(HermesMessage.hermesMessage(TOPIC, CONTENT).build()).join()

        then: //initial attempt
            1 * scheduler.schedule(_, 0, TimeUnit.MILLISECONDS)

        then: //followed  by 2 retries
            2 * scheduler.schedule(_, _, _)
    }

    def "should retry immediately when retry sleep is zero"() {
        given:
        ScheduledExecutorService scheduler = Spy(ScheduledThreadPoolExecutor, constructorArgs: [1])
            HermesClient client = hermesClient({uri, msg -> statusFuture(500)})
                    .withRetries(2)
                    .withRetrySleep(0)
                    .withScheduler(scheduler)
                    .build()

        when:
            client.publish(HermesMessage.hermesMessage(TOPIC, CONTENT).build()).join()

        then: //initial attepmt and 2 retries
            3 * scheduler.schedule(_, 0, _)
    }